docs/CSL-1925.md

```diff
@@ -0,0 +1,256 @@
+
+# Problem Definition
+
+An address in Cardano is always derived from a root key using BIP32 style derivation
+except it uses a different elliptic curve than Bitcoin (Ed25519). 
+
+The problem initially raised concerns exchanges which have only one account per root
+key and thus 2^31 possible addresses. In this scenario according to the birthday
+paradox there is a non negligible chance of a collision occuring.
+
+The first child (or branch) of the root key is derived according to the _account index_
+which by default is 2^31 or in binary `10000000000000000000000000000000` where the
+first bit represents the hardened state. As a BIP32 path this can be represented as 
+`m/aH`. The second child is derived according to an _address index_.
+
+Both the account and address indices fall within the range 2^31 in terms of how many
+addresses can be derived.
+
+In the current scheme the address size is 76 bytes, where each index takes up 5 bytes.
+
+# Generation Strategies
+
+There are currently two ways of generating a unique address, as depicted by the following
+ADT.
+
+```haskell
+data GenerationMode 
+  = RandomMode
+  -- ^ Generates an address where the second index is random and hardened.
+  | DeterministicMode Word32 
+  -- ^ Generates an address where the second index is parameterised.
+```
+
+The difference between `RandomMode` and `DeterministicMode` is that in `DeterministicMode` the
+application / dependent code or user must supply the seed used to generate the index instead 
+of it being generated randomly. `DeterministicMode` is there in order to be able to recover 
+the address at a later date.
+
+The wallet uses a function called `genUniqueAddress` to generate addresses which in turn
+uses the function shown below named `generateUnique`. In RandomMode this function tries 
+to ensure that the index used is unique according to an arbitrary predicate `isDuplicate`. 
+For addresses this predicate is called `doesWAddressExist` which checks storage for a 
+pre-existing address.
+
+In terms of BIP32 paths this extra index forms the `bH` in the derivation path `m/aH/bH` 
```

Comment:
Here is an idea: as different sources uses different conventions for the hardened/non-hardened keys (I have seen lots of them using apostrophes, like in `m/10'/20'/30'`), should we mention we are sticking to the same way [BIP-32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#test-vectors) labels them? probably not necessary, but if we omit the link, I would argue it would be good to have the explanation of what `h` stands for at the beginning of the writeup, as we introduce it fairly soon (on line 14, to be precise)

Reply:
Yes this makes sense, I will stick to BIP-32 labelling and change the document accordingly to reflect this.